File: rover_middleware/Rover.Domain/Profiles/ProfileModels.cs
namespace Rover.Domain.Profiles;

public sealed class GuestProfile
{
    public GuestProfile(Guid profileId, string installationId, DateTimeOffset createdAtUtc, DateTimeOffset updatedAtUtc, UserPreferences preferences)
    {
        ProfileId = profileId;
        InstallationId = installationId;
        CreatedAtUtc = createdAtUtc;
        UpdatedAtUtc = updatedAtUtc;
        Preferences = preferences;
    }

    public Guid ProfileId { get; }
    public string InstallationId { get; }
    public DateTimeOffset CreatedAtUtc { get; }
    public DateTimeOffset UpdatedAtUtc { get; private set; }
    public UserPreferences Preferences { get; private set; }
    public List<SavedDiscovery> SavedDiscoveries { get; } = new();
    public List<PreferenceSignal> PreferenceSignals { get; } = new();
    public List<LearnedPreference> LearnedPreferences { get; } = new();
    public List<StoryInteraction> StoryInteractions { get; } = new();
    public uint Version { get; private set; }

    public void UpdatePreferences(UserPreferences preferences, DateTimeOffset updatedAtUtc)
    {
        Preferences = preferences;
        UpdatedAtUtc = updatedAtUtc;
        Version++;
    }

    public void SaveDiscovery(SavedDiscovery discovery)
    {
        if (SavedDiscoveries.Any(item => item.DiscoveryId.Equals(discovery.DiscoveryId, StringComparison.OrdinalIgnoreCase)))
        {
            return;
        }

        SavedDiscoveries.Add(discovery);
        Version++;
    }

    public void RemoveDiscovery(string discoveryId)
    {
        SavedDiscoveries.RemoveAll(item => item.DiscoveryId.Equals(discoveryId, StringComparison.OrdinalIgnoreCase));
        Version++;
    }

    public void RecordSignal(PreferenceSignal signal)
    {
        if (!Preferences.ImproveRecommendations)
        {
            return;
        }

        PreferenceSignals.Add(signal);
        var learned = LearnedPreferences.FirstOrDefault(item => item.Topic.Equals(signal.Topic, StringComparison.OrdinalIgnoreCase));
        if (learned is null)
        {
            LearnedPreferences.Add(new LearnedPreference(signal.Topic, signal.Weight, signal.Reason, signal.CreatedAtUtc));
            Version++;
            return;
        }

        learned.Apply(signal);
        Version++;
    }

    public void RemoveLearnedPreference(string topic)
    {
        LearnedPreferences.RemoveAll(item => item.Topic.Equals(topic, StringComparison.OrdinalIgnoreCase));
        PreferenceSignals.RemoveAll(item => item.Topic.Equals(topic, StringComparison.OrdinalIgnoreCase));
        Version++;
    }

    public void ResetLearning()
    {
        LearnedPreferences.Clear();
        PreferenceSignals.Clear();
        StoryInteractions.Clear();
        Version++;
    }

    public bool RecordStoryInteraction(StoryInteraction interaction, int maximumEvents, DateTimeOffset oldestAllowedUtc)
    {
        if (!Preferences.ImproveRecommendations
            || StoryInteractions.Any(item => item.EventId.Equals(interaction.EventId, StringComparison.OrdinalIgnoreCase)))
        {
            return false;
        }

        StoryInteractions.RemoveAll(item => item.OccurredAtUtc < oldestAllowedUtc);
        StoryInteractions.Add(interaction);
        var overflow = StoryInteractions.Count - Math.Max(1, maximumEvents);
        if (overflow > 0)
        {
            StoryInteractions.RemoveRange(0, overflow);
        }
        Version++;
        return true;
    }
}

public sealed record UserPreferences(
    IReadOnlyList<string> Interests,
    string WalkingPace,
    IReadOnlyList<string> AccessibilityNeeds,
    string DistanceUnits,
    bool DirectionVoiceEnabled,
    bool NarrationEnabled,
    double SpeechRate,
    string PreferredNarrationLength,
    bool PremiumVoiceEnabled,
    bool AskRoverVoiceEnabled,
    bool AutoPlayNarrationOnArrival,
    bool ResumeNarrationAfterNavigation,
    bool DeviceVoiceFallbackEnabled,
    bool SaveWalkHistory,
    bool ImproveRecommendations,
    string LocationRetention)
{
    public string StoryDensity { get; init; } = "Highlights";
    public IReadOnlyList<string> ExcludedStoryCategories { get; init; } = Array.Empty<string>();

    public static UserPreferences Default => new(
        Array.Empty<string>(),
        "Standard",
        Array.Empty<string>(),
        "miles",
        true,
        true,
        0.48,
        "Balanced",
        true,
        true,
        true,
        true,
        true,
        true,
        false,
        "WalkRestorationOnly");
}

public sealed record SavedDiscovery(Guid SavedDiscoveryId, string DiscoveryId, string Name, string Category, DateTimeOffset SavedAtUtc, string Source);
public sealed record PreferenceSignal(Guid PreferenceSignalId, string Topic, int Weight, string Reason, DateTimeOffset CreatedAtUtc);

public enum StoryInteractionKind
{
    Offered,
    Started,
    Completed,
    Skipped,
    Interrupted,
    Replayed,
    TellMore,
    Dismissed
}

public sealed record StoryInteraction(
    string EventId,
    string StoryId,
    string Category,
    StoryInteractionKind Kind,
    DateTimeOffset OccurredAtUtc);

public sealed class LearnedPreference
{
    public LearnedPreference(string topic, int score, string reason, DateTimeOffset updatedAtUtc)
    {
        Topic = topic;
        Score = Math.Clamp(score, -10, 10);
        Reason = reason;
        UpdatedAtUtc = updatedAtUtc;
    }

    public string Topic { get; }
    public int Score { get; private set; }
    public string Reason { get; private set; }
    public DateTimeOffset UpdatedAtUtc { get; private set; }

    public void Apply(PreferenceSignal signal)
    {
        Score = Math.Clamp(Score + Math.Clamp(signal.Weight, -3, 3), -10, 10);
        Reason = signal.Reason;
        UpdatedAtUtc = signal.CreatedAtUtc;
    }
}

File: rover_flutter/lib/src/api/profile_models.dart
class GuestProfileRequest {
  const GuestProfileRequest({required this.installationId});

  final String installationId;

  Map<String, Object?> toJson() => {'installationId': installationId};
}

class UpdateProfilePreferencesRequest {
  const UpdateProfilePreferencesRequest({
    this.interests,
    this.walkingPace,
    this.accessibilityNeeds,
    this.distanceUnits,
    this.directionVoiceEnabled,
    this.narrationEnabled,
    this.speechRate,
    this.preferredNarrationLength,
    this.storyDensity,
    this.excludedStoryCategories,
    this.premiumVoiceEnabled,
    this.askRoverVoiceEnabled,
    this.autoPlayNarrationOnArrival,
    this.resumeNarrationAfterNavigation,
    this.deviceVoiceFallbackEnabled,
    this.saveWalkHistory,
    this.improveRecommendations,
    this.locationRetention,
  });

  final List<String>? interests;
  final String? walkingPace;
  final List<String>? accessibilityNeeds;
  final String? distanceUnits;
  final bool? directionVoiceEnabled;
  final bool? narrationEnabled;
  final double? speechRate;
  final String? preferredNarrationLength;
  final String? storyDensity;
  final List<String>? excludedStoryCategories;
  final bool? premiumVoiceEnabled;
  final bool? askRoverVoiceEnabled;
  final bool? autoPlayNarrationOnArrival;
  final bool? resumeNarrationAfterNavigation;
  final bool? deviceVoiceFallbackEnabled;
  final bool? saveWalkHistory;
  final bool? improveRecommendations;
  final String? locationRetention;

  Map<String, Object?> toJson() {
    return {
      'interests': interests,
      'walkingPace': walkingPace,
      'accessibilityNeeds': accessibilityNeeds,
      'distanceUnits': distanceUnits,
      'directionVoiceEnabled': directionVoiceEnabled,
      'narrationEnabled': narrationEnabled,
      'speechRate': speechRate,
      'preferredNarrationLength': preferredNarrationLength,
      'storyDensity': storyDensity,
      'excludedStoryCategories': excludedStoryCategories,
      'premiumVoiceEnabled': premiumVoiceEnabled,
      'askRoverVoiceEnabled': askRoverVoiceEnabled,
      'autoPlayNarrationOnArrival': autoPlayNarrationOnArrival,
      'resumeNarrationAfterNavigation': resumeNarrationAfterNavigation,
      'deviceVoiceFallbackEnabled': deviceVoiceFallbackEnabled,
      'saveWalkHistory': saveWalkHistory,
      'improveRecommendations': improveRecommendations,
      'locationRetention': locationRetention,
    };
  }
}

class SaveDiscoveryRequest {
  const SaveDiscoveryRequest({
    required this.discoveryId,
    required this.name,
    required this.category,
    required this.source,
  });

  final String discoveryId;
  final String name;
  final String category;
  final String source;

  Map<String, Object?> toJson() {
    return {
      'discoveryId': discoveryId,
      'name': name,
      'category': category,
      'source': source,
    };
  }
}

class GuestProfile {
  const GuestProfile({
    required this.profileId,
    required this.installationId,
    required this.preferences,
    required this.savedDiscoveries,
    required this.learnedPreferences,
    this.storyInteractionCount = 0,
  });

  final String profileId;
  final String installationId;
  final ProfilePreferences preferences;
  final List<SavedDiscovery> savedDiscoveries;
  final List<LearnedPreference> learnedPreferences;
  final int storyInteractionCount;

  factory GuestProfile.fromJson(Map<String, dynamic> json) {
    return GuestProfile(
      profileId: json['profileId'] as String,
      installationId: json['installationId'] as String,
      preferences: ProfilePreferences.fromJson(
        json['preferences'] as Map<String, dynamic>,
      ),
      savedDiscoveries: _list(json['savedDiscoveries'])
          .map((item) => SavedDiscovery.fromJson(item as Map<String, dynamic>))
          .toList(),
      learnedPreferences: _list(json['learnedPreferences'])
          .map(
            (item) => LearnedPreference.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
      storyInteractionCount: json['storyInteractionCount'] as int? ?? 0,
    );
  }
}

class ProfilePreferences {
  const ProfilePreferences({
    required this.interests,
    required this.walkingPace,
    required this.accessibilityNeeds,
    required this.distanceUnits,
    required this.directionVoiceEnabled,
    required this.narrationEnabled,
    required this.speechRate,
    required this.preferredNarrationLength,
    required this.storyDensity,
    required this.excludedStoryCategories,
    required this.premiumVoiceEnabled,
    required this.askRoverVoiceEnabled,
    required this.autoPlayNarrationOnArrival,
    required this.resumeNarrationAfterNavigation,
    required this.deviceVoiceFallbackEnabled,
    required this.saveWalkHistory,
    required this.improveRecommendations,
    required this.locationRetention,
  });

  final List<String> interests;
  final String walkingPace;
  final List<String> accessibilityNeeds;
  final String distanceUnits;
  final bool directionVoiceEnabled;
  final bool narrationEnabled;
  final double speechRate;
  final String preferredNarrationLength;
  final String storyDensity;
  final List<String> excludedStoryCategories;
  final bool premiumVoiceEnabled;
  final bool askRoverVoiceEnabled;
  final bool autoPlayNarrationOnArrival;
  final bool resumeNarrationAfterNavigation;
  final bool deviceVoiceFallbackEnabled;
  final bool saveWalkHistory;
  final bool improveRecommendations;
  final String locationRetention;

  factory ProfilePreferences.fromJson(Map<String, dynamic> json) {
    return ProfilePreferences(
      interests: _stringList(json['interests']),
      walkingPace: json['walkingPace'] as String,
      accessibilityNeeds: _stringList(json['accessibilityNeeds']),
      distanceUnits: json['distanceUnits'] as String,
      directionVoiceEnabled: json['directionVoiceEnabled'] as bool,
      narrationEnabled: json['narrationEnabled'] as bool,
      speechRate: (json['speechRate'] as num).toDouble(),
      preferredNarrationLength: json['preferredNarrationLength'] as String,
      storyDensity: json['storyDensity'] as String? ?? 'Highlights',
      excludedStoryCategories: _stringList(json['excludedStoryCategories']),
      premiumVoiceEnabled: json['premiumVoiceEnabled'] as bool? ?? true,
      askRoverVoiceEnabled: json['askRoverVoiceEnabled'] as bool? ?? true,
      autoPlayNarrationOnArrival:
          json['autoPlayNarrationOnArrival'] as bool? ?? true,
      resumeNarrationAfterNavigation:
          json['resumeNarrationAfterNavigation'] as bool? ?? true,
      deviceVoiceFallbackEnabled:
          json['deviceVoiceFallbackEnabled'] as bool? ?? true,
      saveWalkHistory: json['saveWalkHistory'] as bool,
      improveRecommendations: json['improveRecommendations'] as bool,
      locationRetention: json['locationRetention'] as String,
    );
  }
}

class SavedDiscovery {
  const SavedDiscovery({
    required this.discoveryId,
    required this.name,
    required this.category,
    required this.savedAtUtc,
    required this.source,
  });

  final String discoveryId;
  final String name;
  final String category;
  final DateTime savedAtUtc;
  final String source;

  factory SavedDiscovery.fromJson(Map<String, dynamic> json) {
    return SavedDiscovery(
      discoveryId: json['discoveryId'] as String,
      name: json['name'] as String,
      category: json['category'] as String,
      savedAtUtc: DateTime.parse(json['savedAtUtc'] as String),
      source: json['source'] as String,
    );
  }
}

class LearnedPreference {
  const LearnedPreference({
    required this.topic,
    required this.score,
    required this.reason,
  });

  final String topic;
  final int score;
  final String reason;

  factory LearnedPreference.fromJson(Map<String, dynamic> json) {
    return LearnedPreference(
      topic: json['topic'] as String,
      score: json['score'] as int,
      reason: json['reason'] as String,
    );
  }
}

List<Object?> _list(Object? value) => value is List ? value : const [];

List<String> _stringList(Object? value) {
  return value is List
      ? value.map((item) => item.toString()).toList()
      : const [];
}
